File: Services/PathfindingService/Models/AABox.cs
namespace VMAP
{
    /// <summary>
    /// Axis-Aligned Bounding Box with minimum and maximum corners.
    /// </summary>
    public struct AABox
    {
        public Vector3 Min;
        public Vector3 Max;

        public AABox(Vector3 min, Vector3 max)
        {
            Min = min;
            Max = max;
        }

        /// <summary>Returns a special AABox representing an empty or zero-size box.</summary>
        public static AABox Zero => new AABox(Vector3.Zero, Vector3.Zero);

        /// <summary>Check if a point lies within this bounding box (inclusive on boundaries).</summary>
        public bool Contains(Vector3 point)
        {
            return (point.x >= Min.x && point.x <= Max.x) &&
                   (point.y >= Min.y && point.y <= Max.y) &&
                   (point.z >= Min.z && point.z <= Max.z);
        }

        /// <summary>Merge this box with a point, expanding it to include the point.</summary>
        public void Merge(Vector3 point)
        {
            Min = Min.Min(point);
            Max = Max.Max(point);
        }

        // Offset the box by a vector (translates both min and max)
        public static AABox operator +(AABox box, Vector3 vec)
        {
            return new AABox(box.Min + vec, box.Max + vec);
        }
        /// <summary>Ray–AABB clip; returns false on miss, true on hit and updates <c>tMin/tMax</c>.</summary>
        public readonly bool Intersects(in Ray r, ref float tMin, ref float tMax)
        {
            Vector3 org = r.Origin, dir = r.Direction;
            Vector3 inv = r.invDirection();

            for (int ax = 0; ax < 3; ++ax)
            {
                float o = org[ax], dInv = inv[ax], mn = Min[ax], mx = Max[ax];
                float t1 = (mn - o) * dInv;
                float t2 = (mx - o) * dInv;
                if (t1 > t2) (t1, t2) = (t2, t1);
                tMin = MathF.Max(tMin, t1);
                tMax = MathF.Min(tMax, t2);
                if (tMin > tMax)
                    return false;
            }
            return true;
        }
        public override string ToString() => $"Box[Min={Min}, Max={Max}]";

        // Equality and hashing to allow comparisons
        public override bool Equals(object? obj)
        {
            if (obj is AABox other)
                return Min == other.Min && Max == other.Max;
            return false;
        }
        public override int GetHashCode() => HashCode.Combine(Min, Max);
        public static bool operator ==(AABox a, AABox b) => a.Equals(b);
        public static bool operator !=(AABox a, AABox b) => !a.Equals(b);
    }
}

File: Exports/Navigation/AABox.cpp
// AABox.cpp
#include "AABox.h"

namespace G3D
{
    AABox::AABox()
        : m_low(inf(), inf(), inf()),
        m_high(-inf(), -inf(), -inf())
    {
    }

    AABox::AABox(const Vector3& lo, const Vector3& hi)
        : m_low(lo), m_high(hi)
    {
    }

    void AABox::set(const Vector3& lo, const Vector3& hi)
    {
        m_low = lo;
        m_high = hi;
    }

    void AABox::merge(const AABox& box)
    {
        m_low = m_low.min(box.m_low);
        m_high = m_high.max(box.m_high);
    }

    void AABox::merge(const Vector3& v)
    {
        m_low = m_low.min(v);
        m_high = m_high.max(v);
    }

    bool AABox::contains(const Vector3& p) const
    {
        return p.x >= m_low.x && p.x <= m_high.x &&
            p.y >= m_low.y && p.y <= m_high.y &&
            p.z >= m_low.z && p.z <= m_high.z;
    }

    Vector3 AABox::corner(int index) const
    {
        Vector3 v;
        switch (index)
        {
        case 0: v = Vector3(m_low.x, m_low.y, m_high.z); break;
        case 1: v = Vector3(m_high.x, m_low.y, m_high.z); break;
        case 2: v = Vector3(m_high.x, m_high.y, m_high.z); break;
        case 3: v = Vector3(m_low.x, m_high.y, m_high.z); break;
        case 4: v = Vector3(m_low.x, m_low.y, m_low.z);  break;
        case 5: v = Vector3(m_high.x, m_low.y, m_low.z);  break;
        case 6: v = Vector3(m_high.x, m_high.y, m_low.z);  break;
        case 7: v = Vector3(m_low.x, m_high.y, m_low.z);  break;
        default:
            // Invalid corner index
            return Vector3::zero();
        }
        return v;
    }

    bool AABox::intersects(const AABox& other) const
    {
        // Must be overlap along all three axes
        for (int a = 0; a < 3; ++a) {
            if ((m_low[a] > other.m_high[a]) || (m_high[a] < other.m_low[a])) {
                return false;
            }
        }
        return true;
    }

    bool AABox::operator==(const AABox& b) const
    {
        return m_low == b.m_low && m_high == b.m_high;
    }
}